---
title: Concordance ADSP vs. Younkin database
author: "SteveYounkin"
date: "November 17, 2015"
output: html_document
---

##Analysis Plan

Identify Mayo subjects included in ADSP WES dataset. Identify ADSP SNPs genotyped independently at Mayo. Evaluate the concordance in the calls made in ADSP with those made independently by Mayo. Will use Joseph's ADSP repo at GitHub to provide version control, will code reproducibly using CompareADSPvsYdb.Rmd.

###What each participant needs:
1. RStudio installed 
2. GitHub installed
3. Local clone of the repo at https://github.com/StevenGYounkin/SORCS2
    + SORCS2_Genetics.Rmd, the file we are generating together, is in this repo
        + All files used or generated by SORCS2_Genetics.Rmd are in this repo
4. R packages installed
    + dplyr
    + data.table
5. PLINK installed and set up so it can be called by R with system("plink......")
6. Perl installed and set up so it can be called by R with system("   ")

###Useful documentation:
1. GitHub
    + https://training.github.com/kit/downloads/github-git-cheat-sheet
2. R Markdown
    + https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf
3. dplyr and tidyr
    + https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf
4. PLINK
    + http://pngu.mgh.harvard.edu/~purcell/plink/
    
###Responsibility for Specific Analysis Steps:
1. Steve will provide genotypes from the Younkin database and do the analysis 
2. Joseph will provide genotypes from the ADSP and assist Steve to identify subjects and SNPs
  
###Steve's Content


```{r, echo=T, eval=T}

        #load requisite packages       
        library(data.table)
        suppressWarnings(suppressMessages(library(dplyr)))
        
        #Identify subjects by reading in the key file created by Joseph,and Jez.
        
        adsp <- read.table("ADSP_sample_info_0M_1F.csv", sep=',', stringsAsFactors = FALSE, header=T) #File from Joseph
        #str(adsp)
        #adsp <- distinct(adsp)
        #str(adsp)
        #x <- as.factor(adsp$StudyName)
        #print (levels(x))
        #print.data.frame (adsp %>% group_by(StudyName) %>% summarize (n=n()))
        #lgen <- filter(lgen, Call %in% call)
        #adspm <- filter(adsp, StudyName=="Mayo Clinic")
        #str(adspm)
        #The file from Joseph ("ADSP_sample_info_0M_1F.csv") has 349 distinct rows with StudyName = "Mayo Clinic"
        
        link3 <- read.table("ADSPSubjectsWithCovariatesForSteve_Jez_11172015.csv", sep=',', stringsAsFactors = FALSE, header=T) #File from Jez
        str(link3)
        link3 <- distinct(link3)
        str(link3)
        both1 <- link3 %>% select(SMID:AD,SRR,SubKey_Younkin:CaseControlDxAge)
        str(both1)
        both2 <- both1 %>% select(StudyName,ADSP_SM_ID,SubKey_Younkin,SubjectId_NETdb,AD,Autopsy,Sex,Age,APOE,Braak,YounkinSeries:CaseControlDxAge)
        str(both2)
        ck1 <- both2 %>% group_by(StudyName,Sex,Gender) %>% summarize(n=n())
        print(ck1)
        ck2 <- both2 %>% group_by(StudyName,APOE,ConsensusApoE) %>% summarize(n=n())
        print(ck2)
        ck3 <- both2 %>% group_by(StudyName,AD,Dx1,Dx2) %>% summarize(n=n())
        print(ck3)
        ck4 <- both2 %>% group_by(StudyName,Autopsy,YounkinSeries,SeriesList) %>% summarize(n=n())
        print(ck4)
        
        #These checks confirm that: 
        #(1) Mayo Clinic in study name identifies the 349 Mayo subjects in the ADSP dataset
        #(2) The number at the end of ADSP_SM_ID corresponds to SubKey in the Younkin database 
        
        #Read in Ydb call on SNPs in ABCA7, TLR4, TLR5, and SORCS2
        ydb <- read.table("4GenesWithExonicSNPgenotypes.csv", sep=',', stringsAsFactors = FALSE, header=T)
        str(ydb)
        var <- ydb %>% select(VariantID:position)
        var <- distinct(var)
        str(var)
        print.data.frame(var)
        
        msub <- ydb %>% select(SubKey:ApoE)
        str(msub)
        msub <- distinct(msub)
        str(msub)
        both1 <- both1 %>% rename(SubKey=SubKey_Younkin)
        b1 <- inner_join(msub,both1, by="SubKey")
        str(b1)  
        #All 349 Mayo subjects in adsp are in 4Genes file
        
        #Join
        bg1 <- inner_join(b1,ydb, by="SubKey")
        str(bg1)
        bg1sub <- bg1 %>% select(SubKey,SRR,rs,Call,BestCall)
        str(bg1sub)
        
        adb <- read.table("ADSP_3000_flat_file.txt", stringsAsFactors = FALSE, header=T)
        str(adb)
        
        c1 <- inner_join(bg1sub,adb,by=c("SRR" = "Subject_ID", "rs" ="SNP_ID"))
        str(c1)
        c2 <- c1 %>% group_by(rs,Call.x, Call.y) %>% summarize(n=n())
        print.data.frame(c2)
        c3 <- c1 %>% group_by (rs) %>% summarize(n=n())
        print.data.frame(c3)
        c4 <- c1 %>% group_by(Call.x, Call.y) %>% summarize(n=n())
        print(c4)
            
        #bg2 <- bg1 %>% select(VariantID:Call,BestCall)
        #bg2 <- distinct(bg2)
        #print.data.frame(bg2)
        #rsct <- bg2 %>% group_by(VariantID,VariantName,rs,Call) %>% summarize(n=n())
        #print.data.frame (rsct)
        
       
        

        
```
        
        
```{r, eval=F}
        
        #x <- as.factor(adsp$StudyName)
        #print (levels(x))
        #print.data.frame (adsp %>% group_by(StudyName) %>% summarize (n=n()))
        #lgen <- filter(lgen, Call %in% call)
        #adspm <- filter(adsp, StudyName=="Mayo Clinic")
        #str(adspm)
        #The file from Joseph ("ADSP_sample_info_0M_1F.csv") has 349 distinct rows with StudyName = "Mayo Clinic"
        
        
        link1 <- read.table("qryADGC_05-07-2013_WORKFLOW.csv", sep=',', stringsAsFactors = FALSE, header=T)
        str(link1)
        
        comb1 <- inner_join(link1,adsp,by="ADSPID")
        head(comb1)
        str(comb1)
        link2 <- read.table("qryADGC_215_chosen-pulled_07-02-2013_For-Sarah.csv", sep=',', stringsAsFactors = FALSE, header=T)
        #str(link2)
        link2 <- link2 %>% select(ADSPID,SubKey,SubjectIDType,SubjectIDValue,Dx,Series,Gend,Sex,Plate)
        str(link2)
        comb2 <- inner_join(link2,adsp,by="ADSPID")
        head(comb2)
        str(comb2)
        comb3 <- comb1 %>% select(SubKey,ADSPID,Gender,Age,APOE,Autopsy,Braak,PrevAD,IncAD,AD)
        comb4 <- comb2 %>% select(SubKey,ADSPID,Gender,Age,APOE,Autopsy,Braak,PrevAD,IncAD,AD)
        comb5 <- distinct(bind_rows (comb3,comb4))
        str(comb5); head(comb5)
        #print.data.frame (comb1 %>% group_by (TYPE,PrevAD,IncAD,AD) %>% summarize(n=n()))
        str(link1);str(link2)
        comb6 <- inner_join(link1,link2,by="ADSPID")
        str(comb6)
```

```{r, echo=F, eval=F}
        str(data);head(data); tail(data)

        #To debug, it was convenient to provide the output called by the function here
        #output <- "SORCS2.CM"
        
        #Create output files
        out1 <- paste(output,sep="",".map")
        out2 <- paste(output,sep="",".fam")
        out3 <- paste(output,sep="","_cov.txt")
        out4 <- paste(output,sep="",".lgen")
        out5 <- paste(output,sep="",".plink.csv")
        #print(out1);print(out2);print(out3);print(out4);print(out5)

        #Generate PLINK map file by copying and renaming "SORCS2_YdB.map"
        map <- read.table("SORCS2_YdB.map", stringsAsFactors = FALSE, header=F)
        #print(map,row.names=F, col.names=F)
        write.table(map,out1,sep='\t',quote=FALSE,row.names=F,col.names=F)

        #In his flat file, Chris encoded each SNP as a separate column with 0,1,2 coding
        #There'is additional wrangling to do to produce map, fam, cov, and lgen files

        #Add columns that are not in Chris file and rename as needed
        data <- mutate(data, PID=0, MID=0, rs="rs", A1=0, A2=0)
        data <- rename (data,IID=SubKey, AFF=Dx2, JS=JS_SIB, NCRAD=INDIANA, Age=DxAge, APOE4dose=APOE4, APOE2dose=APOE2)
        #str(data)

        #Recode as needed to conform to PLINK cov and fam conventions
        data[data$AFF == 1,][, "AFF"] <- 2
        data[data$AFF == 0,][, "AFF"] <- 1

        data[data$SEX == 1,][, "SEX"] <- 2
        data[data$SEX == 0,][, "SEX"] <- 1

        #Check coding for PLINK
        #print(data %>% group_by(AFF,SEX,APOE4dose, APOE2dose,A1,A2) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(A1,A2) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(rs,AFF) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(SEX) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(APOE4dose,APOE2dose) %>% summarize(n=n()))      
        #print.data.frame(y)

        #Create PLINK fam file
        fam <- select(data,FID,IID,PID,MID,SEX,AFF)
        fam <- distinct(fam)
        #str(fam); head(fam); tail(fam)
        write.table(fam,out2,sep='\t',quote=FALSE,row.names=F,col.names=F)
        #SORCS2fam=read.table("SORCS2.fam", header = F)
        #str(SORCS2fam);head(SORCS2fam);tail(SORCS2fam)

        #Create PLINK cov file; 
        #Not sure why, but errors sometimes occurred in r when using out3.cov" as output using write.table
        #So I used "out3_cov.txt" thereby avoiding this problem
        cov <- select(data,FID,IID,JS,RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose)
        cov <- distinct(cov)
        #str(cov)
        write.table(cov,out3,sep='\t',quote=FALSE,row.names=F,col.names=T)
        #SORCS2cov=read.table, out3, header = T)
        #str(SORCS2cov);head(SORCS2cov); tail(SORCS2cov)

        #Create PLINK lgen file, this could have been done better by creating a function to add each
        # to the bottom of an lgen file, and calling the function successively. I copied and revised the code.
        
        #Create lgen file with rs34058821 genotypes
        lgen <- select(data,FID,IID,SORCS2_rs34058821_GA,rs,A1,A2)
        #str(lgen)
        lgen <- rename(lgen, Call=SORCS2_rs34058821_GA)
        #str(lgen)
        lgen <- mutate(lgen, rs="rs34058821")
        #str(lgen)
        # Keep only calls of 0,1,or 2 so there is no missing data
        call <- c(0,1,2)
        lgen <- filter(lgen, Call %in% call)
        #print.data.frame(lgen %>% group_by(rs,Call) %>% summarize(n()))

        #Create lgen file with rs35935435 genotypes
        x <- select(data,FID,IID,SORCS2_rs35935435_CA,rs,A1,A2)
        #str(x)
        x <- rename(x, Call=SORCS2_rs35935435_CA)
        #str(x)
        x <- mutate(x, rs="rs35935435")
        #str(x)
        # Keep only calls of 0,1,or 2 so there is no missing data
        call <- c(0,1,2)
        x <- filter(x, Call %in% call) 
        #print.data.frame(x %>% group_by(rs,Call) %>% summarize(n()))

        #Add rs35935435 genotypes in lgen format to the bottom of lgen 
        lgen <- bind_rows(lgen,x)
        #print.data.frame(lgen %>% group_by(rs,Call) %>% summarize(n()))

        #Create lgen file with rs16840892 genotypes
        x <- select(data,FID,IID,SORCS2_rs16840892_CT,rs,A1,A2)
        #str(x)
        x <- rename(x, Call=SORCS2_rs16840892_CT)
        #str(x)
        x <- mutate(x, rs="rs16840892")
        #str(x)
        # Keep only calls of 0,1,or 2 so there is no missing data
        call <- c(0,1,2)
        x <- filter(x, Call %in% call)
        #print.data.frame(x %>% group_by(rs,Call) %>% summarize(n()))

        #Add rs16840892 genotypes in lgen format to the bottom of lgen
        lgen <- bind_rows(lgen,x)
        #print.data.frame(lgen %>% group_by(rs,Call) %>% summarize(n()))

        #Create lgen file with rs6816604 genotypes
        x <- select(data,FID,IID,SORCS2_rs6816604_CT,rs,A1,A2)
        #str(x)
        x <- rename(x, Call=SORCS2_rs6816604_CT)
        #str(x)
        x <- mutate(x, rs="rs6816604")
        #str(x)
        call <- c(0,1,2)
        x <- filter(x, Call %in% call)
        #print.data.frame(x %>% group_by(rs,Call) %>% summarize(n()))

        #Add rs6816604 genotypes in lgen format to the bottom of lgen
        lgen <- bind_rows(lgen,x)
        #print.data.frame(lgen %>% group_by(rs,Call) %>% summarize(n()))

        #Separate the alleles as required by PLINK
        lgen[lgen$Call == 0,][, "A1"] <- 1
        lgen[lgen$Call == 1,][, "A1"] <- 1
        lgen[lgen$Call == 2,][, "A1"] <- 2

        lgen[lgen$Call == 0,][, "A2"] <- 1
        lgen[lgen$Call == 1,][, "A2"] <- 2
        lgen[lgen$Call == 2,][, "A2"] <- 2
        #For debugging use commands like the one below to check format
        #print.data.frame(lgen %>% group_by(rs,Call,A1,A2) %>% summarize(n()))

        #Select columns for lgen file and write it out
        lgen <- select(lgen,FID,IID,rs,A1,A2)
        #str(lgen)
        write.table(lgen,out4,sep='\t',quote=FALSE,row.names=F,col.names=F)
        #SORCS2lgen=read.table("SORCS2.lgen", header = F)
        #str(SORCS2lgen)

        #Analyse the files in PLINK, format, and print the output. The analysis here is the same as that 
        #in the script for the adsp2plink function below, where it is better annotated.
        
        assoc <- paste("plink",sep=" ","--lfile", output, "--assoc", "--ci 0.95", "--out", output)
        logistic <- paste("plink",sep=" ","--lfile", output, "--logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --out", output)
        haplog <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --out", output)

        assocfile <- paste(output,sep="",".assoc")
        logisfile <- paste(output,sep="",".assoc.logistic")
        haplogfile <- paste(output,sep="",".assoc.hap.logistic")

        system(assoc)
        system(logistic)
        system(haplog)

        f1 <- read.table(assocfile, header=T)
        #print(f1)
        f1 <- f1 %>% select(CHR,SNP,BP,F_A, F_U, OR, P) %>% rename(FREQ_AD=F_A, FREQ_CON=F_U, OR_assoc=OR, P_assoc=P)
        print(f1,digits=2,row.names=F)
        print("", quote=F)
        #head(f1)

        f1 <- f1 %>% select(CHR,SNP,BP,FREQ_AD,FREQ_CON)
        f2 <- read.table(logisfile, header=T)
        #print(f2)
        f2 <- f2 %>% filter (TEST=="ADD") %>% select (SNP,NMISS,OR,L95,U95,P) %>% rename(OR_logis=OR, P_logis=P)
        #head(f2)
        f3<- left_join(f1,f2,by="SNP")
        print(f3,digits=2,row.names=F)
        print("", quote=F)
        write.table(f3,out5,sep=",", quote=FALSE,row.names=F,col.names=T)

        f4 <- read.table(haplogfile, header=T)
        f4 <- f4 %>% select(-NSNP,-NHAP,-STAT) %>% arrange(desc(HAPLOTYPE))
        print(f4,digits=2, row.names=F)
        print("", quote=F)

        hapomni <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --hap-omnibus --out", output)
        system(hapomni)
        f4 <- read.table(haplogfile, header=T)
        f4 <- f4 %>% rename(Pomni=P)
        print(f4,digits=2, row.names=F)
        print("", quote=F)
}

```


```{r, eval=T, echo=F}

        #Create function to read in SQL-derived dataset from Younkin database, 
        #make PLINK files (map, fam, cov.txt, and lgen) and analyze them.

ydb2plink <- function(input="SORCS2_Ydb.csv", output="SORCS2.Ydb") {

        #load requisite packages       
        library(data.table)
        suppressWarnings(suppressMessages(library(dplyr)))

        #To debug, it was convenient to provide the input called by the function here
        #input="SORCS2_Ydb.csv"   
        
        #Read in dataset generated from the Younkin database
        data <- read.table(input, sep=',', stringsAsFactors = FALSE, header=T)

        #Create output files
        
        #To debug, it was convenient to provide the output called by the function here
        #output <- "SORCS2.Ydb"
        
        #Generate names for the PLINK files and the .csv files that will be generated
        out1 <- paste(output,sep="",".map")
        out2 <- paste(output,sep="",".fam")
        out3 <- paste(output,sep="","_cov.txt")
        out4 <- paste(output,sep="",".lgen")
        out5 <- paste(output,sep="",".plink.csv")
        #print(out1);print(out2);print(out3);print(out4);print(out5)

        #Generate PLINK map file
        map <- select(data,chr, rs, position)
        map <- mutate(map,morgans=0)
        map <- rename(map, bp=position)
        map <- select(map,chr,rs,morgans,bp)
        map <- distinct(map)
        #print(map)
        write.table(map,out1,sep='\t',quote=FALSE,row.names=F,col.names=F)

        #Filter, using dplyr package, to include only subjects with: 
        #Dx = AD or CON, Gend = M or F, ApoE = 22,23,24,33,34,or 44 
        #Series = JS,SibPair_White,RS,AUT,Indiana,NW, DxAge >= 65, BestCall = 1
        #Age (at diagnosis for AD/at entry for CON) of 65 or over
        dx <- c("AD","CON")
        sex <- c("M","F")
        apoe <- c(22,23,24,33,34,44)
        series <- c("JS", "RS", "AUT", "Indiana", "SibPair_White", "NW")
        data <- filter(data, Dx %in% dx, Gend %in% sex, ApoE %in% apoe, Series %in% series, BestCall == 1)
        data <-suppressWarnings(mutate(data,Age=as.numeric(DxAge)))
        data <- filter (data, Age >= 65)

        #Create columns for PLINK fam, cov, and lgen files
        data <- mutate(data, AFF=99, A1=99, A2=99, APOE4dose=99, APOE2dose=99, SEX=99, JS=0, RS=0, AUT=0, NCRAD=0,NW=0, FID=0, IID=SubKey, PID=0, MID=0)
        
        #Populate variables, making sure to hew to PLINK convention

        data[data$Dx == "AD",][, "AFF"] <- 2
        data[data$Dx == "CON",][, "AFF"] <- 1

        data[data$Call == 11,][, "A1"] <- 1
        data[data$Call == 12,][, "A1"] <- 1
        data[data$Call == 22,][, "A1"] <- 2

        data[data$Call == 11,][, "A2"] <- 1
        data[data$Call == 12,][, "A2"] <- 2
        data[data$Call == 22,][, "A2"] <- 2

        data[data$Gend == "M",][, "SEX"] <- 1
        data[data$Gend == "F",][, "SEX"] <- 2

        data[data$ApoE == 22,][, "APOE4dose"] <- 0
        data[data$ApoE == 23,][, "APOE4dose"] <- 0
        data[data$ApoE == 24,][, "APOE4dose"] <- 1
        data[data$ApoE == 33,][, "APOE4dose"] <- 0
        data[data$ApoE == 34,][, "APOE4dose"] <- 1
        data[data$ApoE == 44,][, "APOE4dose"] <- 2

        data[data$ApoE == 22,][, "APOE2dose"] <- 2
        data[data$ApoE == 23,][, "APOE2dose"] <- 1
        data[data$ApoE == 24,][, "APOE2dose"] <- 1
        data[data$ApoE == 33,][, "APOE2dose"] <- 0
        data[data$ApoE == 34,][, "APOE2dose"] <- 0
        data[data$ApoE == 44,][, "APOE2dose"] <- 0

        data[data$Series == "JS",][, "JS"] <- 1
        data[data$Series == "SibPair_White",][, "JS"] <- 1
        data[data$Series == "RS",][, "RS"] <- 1
        data[data$Series == "AUT",][, "AUT"] <- 1
        data[data$Series == "Indiana",][, "NCRAD"] <- 1
        data[data$Series == "NW",][, "NW"] <- 1

        #Check coding for PLINK during debugging, using dplyr as exemplified below
        #print(data %>% group_by(rs,AFF,SEX,APOE4dose, APOE2dose,A1,A2) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(rs,A1,A2) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(rs,AFF) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(rs,SEX) %>% summarize(n=n()))
        #print.data.frame(data %>% group_by(APOE4dose,APOE2dose) %>% summarize(n=n()))      
        #print.data.frame(y)

        #Create PLINK fam file, commands commented out were useful during debugging
        fam <- select(data,FID,IID,PID,MID,SEX,AFF)
        fam <- distinct(fam)
        #str(fam); head(fam); tail(fam)
        write.table(fam,out2,sep='\t',quote=FALSE,row.names=F,col.names=F)
        #SORCS2fam=read.table("SORCS2.fam", header = F)
        #str(SORCS2fam);head(SORCS2fam);tail(SORCS2fam)

        #Create PLINK cov file, commands commented out were useful during debugging
        #Not sure why, but errors sometimes occurred in r when using "SORCS2.cov"" as output using write.table
        #So I used "SORCS2cov.txt" thereby avoiding this problem
        cov <- select(data,FID,IID,JS,RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose)
        cov <- distinct(cov)
        #str(cov)
        write.table(cov,out3,sep='\t',quote=FALSE,row.names=F,col.names=T)
        #SORCS2cov=read.table("SORCS2cov.txt", header = T)
        #str(SORCS2cov);head(SORCS2cov); tail(SORCS2cov)

        #Create PLINK lgen file
        lgen <- select(data,FID,IID,rs,A1,A2)
        #str(lgen)
        write.table(lgen,out4,sep='\t',quote=FALSE,row.names=F,col.names=F)
        #SORCS2lgen=read.table("SORCS2.lgen", header = F)
        #str(SORCS2lgen)

        #Analyse the files in PLINK, format, and print the output. The analysis here is the same as that 
        #in the script for the adsp2plink function below, where it is better annotated.        
        
        assoc <- paste("plink",sep=" ","--lfile", output, "--assoc", "--ci 0.95", "--out", output)
        logistic <- paste("plink",sep=" ","--lfile", output, "--logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --out", output)
        haplog <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --out", output)

        assocfile <- paste(output,sep="",".assoc")
        logisfile <- paste(output,sep="",".assoc.logistic")
        haplogfile <- paste(output,sep="",".assoc.hap.logistic")

        system(assoc)
        system(logistic)
        system(haplog)

        f1 <- read.table(assocfile, header=T)
        #print(f1)
        f1 <- f1 %>% select(CHR,SNP,BP,F_A, F_U, OR, P) %>% rename(FREQ_AD=F_A, FREQ_CON=F_U, OR_assoc=OR, P_assoc=P)
        print(f1,digits=2,row.names=F)
        print("", quote=F)
        #head(f1)

        f1 <- f1 %>% select(CHR,SNP,BP,FREQ_AD,FREQ_CON)
        f2 <- read.table(logisfile, header=T)
        #print(f2)
        f2 <- f2 %>% filter (TEST=="ADD") %>% select (SNP,NMISS,OR,L95,U95,P) %>% rename(OR_logis=OR, P_logis=P)
        #head(f2)
        f3<- left_join(f1,f2,by="SNP")
        print(f3,digits=2,row.names=F)
        print("", quote=F)
        write.table(f3,out5,sep=",", quote=FALSE,row.names=F,col.names=T)
        #print("PLINK commands:");print(assoc);print(logistic)

        f4 <- read.table(haplogfile, header=T)
        f4 <- f4 %>% select(-NSNP,-NHAP,-STAT) %>% arrange(desc(HAPLOTYPE))
        print(f4,digits=2, row.names=F)
        print("", quote=F)

        hapomni <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --hap-omnibus --out", output)
        system(hapomni)
        f4 <- read.table(haplogfile, header=T)
        f4 <- f4 %>% rename(Pomni=P)
        print(f4,digits=2, row.names=F)
        print("", quote=F)
}

```

####Function (adsp2plink) to read in and analyze ADSP SORCS2 dataset

The data we have on the 4 common SORCS2 missense variants in our ADSP dataset was captured by Joseph Reddy in a flatfile (SORCS2_ff.txt) as described in his section below. I read in the dataset (SORCS2_ff.txt) from the SORSC2 repo.The following function reads the data in, creates PLINK, files and then analyzes them in PLINK as for the SORCS2 dataset from the Younkin database.

```{r, eval=T, echo=F}
        #Create function that (i) reads in the ADSP dataset generated by Joseph Reddy, (ii) outputs PLINK map, fam, cov.txt, and lgen files,
        #(iii) analyzes in PLINK using --assoc, as well as --logistic, and --hap-logistic with appropriate covariates, 
        #(iv) prints out well organized results
        
adsp2plink <- function(input="SORCS2_ff.txt", output="SORCS2.adsp") {

        #load requisite packages       
        library(data.table)
        suppressWarnings(suppressMessages(library(dplyr)))  
        
        #While debugging it is helpful to set input and output values that will be called by the function
        #It is also useful to employ the various str and print commands that are commented out below
        #input <- "SORCS2_ff.txt"
        #output <- "SORCS2.adsp"        
        #Read in data on 4 common SORCS2 missense variants in the ADSP database        
        
        data <- read.table(input, stringsAsFactors = FALSE, header=T)
        #str(data); head(data); tail(data)

        #Create output files
        out1 <- paste(output,sep="",".map")
        out2 <- paste(output,sep="",".fam")
        out3 <- paste(output,sep="","_cov.txt")
        out4 <- paste(output,sep="",".lgen")
        out5 <- paste(output,sep="",".plink.csv")
        #print(out1);print(out2);print(out3);print(out4);print(out5)

        #Filter, using dplyr, to include only subjects with: 
        #Case.control = 2 or 1 corresponding to AD or CON, Sex = 0 or 1 (M or F), APOE = 22,23,24,33,34,or 44 
        #Series = ADSP, Age >= 65, Call = 11,12, or 22
        library(dplyr)
        dx_1 <- c(2,1,0)
        sex_1 <- c(0,1)
        apoe_1 <- c(22,23,24,33,34,44)
        series_1 <- c("ADSP")
        call_1 <- c(11,12,22)
        data <- filter(data, Case.control %in% dx_1, Sex %in% sex_1, APOE %in% apoe_1, Series %in% series_1, Call %in% call_1)
        data <- filter (data, Age >= 65)
        #str(data);head(data);tail(data)

        #Create PLINK map file
        map <- select(data, SNP_ID, Position)
        map <- mutate(map,morgans=0, chr=4)
        map <- rename(map, bp=Position, rs=SNP_ID)
        map <- select(map,chr,rs,morgans,bp)
        map <- distinct(map)
        snps <- c("rs6816604","rs35935435","rs16840892","rs34058821")
        map <- filter(map,rs %in% snps )
        print(map)
        write.table(map,out1,sep='\t',quote=FALSE,row.names=F,col.names=F)

        #Create variables for PLINK fam, cov, and lgen files
        data <- mutate(data, AFF=99, A1=99, A2=99, APOE4dose=99, APOE2dose=99, SEX=99, ADSP=0, FID=0, IID=as.character(Subject_ID), PID=0, MID=0)

        data[data$Case.control == 0,][, "AFF"] <- 1  #CON=0
        data[data$Case.control == 1,][, "AFF"] <- 2  #AD=1
        data[data$Case.control == 2,][, "AFF"] <- 2  #probable AD=2 are included as AD for PLINK analysis

        data[data$Call == 11,][, "A1"] <- 1  #1 = major allele, 2 = minor allele
        data[data$Call == 12,][, "A1"] <- 1  #split the allele calls as required by PLINK
        data[data$Call == 22,][, "A1"] <- 2  

        data[data$Call == 11,][, "A2"] <- 1
        data[data$Call == 12,][, "A2"] <- 2
        data[data$Call == 22,][, "A2"] <- 2

        data[data$Sex == 0,][, "SEX"] <- 1   #M=0
        data[data$Sex == 1,][, "SEX"] <- 2   #F=1

        data[data$APOE == 22,][, "APOE4dose"] <- 0 #code dose of A4 allele, 0,1,or2
        data[data$APOE == 23,][, "APOE4dose"] <- 0
        data[data$APOE == 24,][, "APOE4dose"] <- 1
        data[data$APOE == 33,][, "APOE4dose"] <- 0
        data[data$APOE == 34,][, "APOE4dose"] <- 1
        data[data$APOE == 44,][, "APOE4dose"] <- 2

        data[data$APOE == 22,][, "APOE2dose"] <- 2 #code dose of A4 allele, 0,1,or2
        data[data$APOE == 23,][, "APOE2dose"] <- 1
        data[data$APOE == 24,][, "APOE2dose"] <- 1
        data[data$APOE == 33,][, "APOE2dose"] <- 0
        data[data$APOE == 34,][, "APOE2dose"] <- 0
        data[data$APOE == 44,][, "APOE2dose"] <- 0

        data[data$Series == "ADSP",][, "ADSP"] <- 1 #1 indicates subject is ADSP, otherwise would be 0
        #For debugging, evaluate whether data has been formatted correctly by filtering and grouping appropriately
        #y <- data %>% filter(rs=="rs6816604",JS==1) %>% group_by(rs,Dx,A1,A2) %>% summarize(n=n())
        #print(y)

        #Create PLINK fam file
        fam <- select(data,FID,IID,PID,MID,SEX,AFF)
        fam <- distinct(fam)
        #str(fam); head(fam); tail(fam)
        write.table(fam,out2,sep='\t',quote=FALSE,row.names=F,col.names=F)
        #SORCS2fam_1=read.table("SORCS2_1.fam", header = F)
        #str(SORCS2fam_1);head(SORCS2fam_1);tail(SORCS2fam_1)

        #Create PLINK cov file; 
        #Not sure why, but errors sometimes occurred in r when using "SORCS2.cov"" as output using write.table
        #So I used "SORCS2cov.txt" thereby avoiding this problem
        cov <- select(data,FID,IID,ADSP,SEX,Age,APOE4dose,APOE2dose)
        cov <- distinct(cov)
        #For debugging make sure structure looks good
        #str(cov)
        write.table(cov,out3,sep='\t',quote=FALSE,row.names=F,col.names=T)
        #For debugging read table back in and make sure it looks right
        #SORCS2cov_1=read.table("SORCS2_1cov.txt", header = T)
        #str(SORCS2cov_1);head(SORCS2cov_1); tail(SORCS2cov_1)

        #Create PLINK lgen file
        lgen <- select(data,FID,IID,SNP_ID,A1,A2)
        #str(lgen)
        write.table(lgen,out4,sep='\t',quote=FALSE,row.names=F,col.names=F)

        #Create command strings for --assoc, --logistic, and --haplogistic
        assoc <- paste("plink",sep=" ","--lfile", output, "--assoc", "--ci 0.95", "--out", output)
        logistic <- paste("plink",sep=" ","--lfile", output, "--logistic", "--covar", out3, "--covar-name SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--out", output)
        haplog <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --out", output)

        #Create names for the files that PLINK generates
        assocfile <- paste(output,sep="",".assoc")
        logisfile <- paste(output,sep="",".assoc.logistic")
        haplogfile <- paste(output,sep="",".assoc.hap.logistic")

        #Run the PLINK analyses
        system(assoc)
        system(logistic)
        system(haplog)

        #Read in, format, and print the results from --assoc
        f1 <- read.table(assocfile, header=T)
        #print(f1)
        f1 <- f1 %>% select(CHR,SNP,BP,F_A, F_U, OR, P) %>% rename(FREQ_AD=F_A, FREQ_CON=F_U, OR_assoc=OR, P_assoc=P)
        print(f1,digits=2,row.names=F)
        print("", quote=F)
        #head(f1)

        #Read in results for --logistic, combine with results from --logistic, format, print
        f1 <- f1 %>% select(CHR,SNP,BP,FREQ_AD,FREQ_CON)
        f2 <- read.table(logisfile, header=T)
        #print(f2)
        f2 <- f2 %>% filter (TEST=="ADD") %>% select (SNP,NMISS,OR,L95,U95,P) %>% rename(OR_logis=OR, P_logis=P)
        #head(f2)
        f3<- left_join(f1,f2,by="SNP")
        print(f3,digits=2,row.names=F)
        print("", quote=F)
        write.table(f3,out5,sep=",", quote=FALSE,row.names=F,col.names=T)

        #Read in results for --hap-logistic, format, print
        f4 <- read.table(haplogfile, header=T)
        f4 <- f4 %>% select(-NSNP,-NHAP,-STAT) %>% arrange(desc(HAPLOTYPE))
        print(f4,digits=2, row.names=F)
        print("", quote=F)

        #Run and read in results for "--hap-logistic --hap-omnibus"", format, print
        hapomni <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --hap-omnibus --out", output)
        system(hapomni)
        f4 <- read.table(haplogfile, header=T)
        f4 <- f4 %>% rename(Pomni=P)
        print(f4,digits=2, row.names=F)
        print("", quote=F)
}
```

####Function (comb2plink) to create combined ydb-adsp dataset and analyse it
```{r, echo=F, eval=T}
comb2plink <- function(output="SORCS2.comb") {

                #load requisite packages       
library(data.table)
suppressWarnings(suppressMessages(library(dplyr)))

        #Create output files
output <- "SORCS2.comb"
out1 <- paste(output,sep="",".map")
out2 <- paste(output,sep="",".fam")
out3 <- paste(output,sep="","_cov.txt")
out4 <- paste(output,sep="",".lgen")
out5 <- paste(output,sep="",".plink.csv") 

        #Generate PLINK map file by copying and renaming "SORCS2.adsp.map"
map <- read.table("SORCS2.adsp.map", header = F)
write.table(map, out1, sep='\t',quote=FALSE,row.names=F,col.names=F)

        #Use dplyr bind_rows to generate combined fam, cov, and lgen files

fam.ydb <- read.table("SORCS2.Ydb.fam", header=F, stringsAsFactors = FALSE)
colnames(fam.ydb) <- c("FID","IID","PID","MID","SEX","AFF")
  #Change IID from number to character for compatiblity with fam.adsp
fam.ydb <- fam.ydb %>% mutate(IID=as.character(IID)) 
#str(fam.ydb)
fam.adsp <- read.table("SORCS2.adsp.fam", header=F, stringsAsFactors = FALSE)
colnames(fam.adsp) <- c("FID","IID","PID","MID","SEX","AFF")
#str(fam.adsp)
fam.comb <- bind_rows(fam.ydb,fam.adsp)
#str(fam.comb)
write.table(fam.comb, out2, sep='\t',quote=FALSE,row.names=F,col.names=F)

cov.ydb <- read.table("SORCS2.Ydb_cov.txt", header=T, stringsAsFactors = FALSE)
cov.ydb <- cov.ydb %>% mutate(IID=as.character(IID))
#str(cov.ydb)
cov.adsp <- read.table("SORCS2.adsp_cov.txt", header=T, stringsAsFactors = FALSE)
#str(cov.adsp)
cov.comb <- bind_rows(cov.ydb,cov.adsp)
#str(cov.comb)
x <- cov.comb %>% group_by(JS,RS,AUT,NCRAD,NW,ADSP) %>% summarise(n())
#print(x)
                #Combined series covariates had NAs that should be 0                
cov.comb[is.na(cov.comb)]<-0
x <- cov.comb %>% group_by(JS,RS,AUT,NCRAD,NW,ADSP) %>% summarise(n())
#print(x)
                # NA replaced by 0 in all series covariates
write.table(cov.comb, out3, sep='\t',quote=FALSE,row.names=F,col.names=T)

lgen.ydb <- read.table("SORCS2.Ydb.lgen", header=F, stringsAsFactors = FALSE)
colnames(lgen.ydb) <- c("FID","IID","rs","A1","A2")
lgen.ydb <- lgen.ydb %>% mutate(IID=as.character(IID))
#str(lgen.ydb)
lgen.adsp <- read.table("SORCS2.adsp.lgen", header=F, stringsAsFactors = FALSE)
colnames(lgen.adsp) <- c("FID","IID","rs","A1","A2")
#str(lgen.adsp)
lgen.comb <- bind_rows(lgen.ydb,lgen.adsp)
#str(lgen.comb)
write.table(lgen.comb, out4, sep='\t',quote=FALSE,row.names=F,col.names=F)

assoc <- paste("plink",sep=" ","--lfile", output, "--assoc", "--ci 0.95", "--out", output)
logistic <- paste("plink",sep=" ","--lfile", output, "--logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--out", output)
haplog <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,ADSP,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --out", output)


assocfile <- paste(output,sep="",".assoc")
logisfile <- paste(output,sep="",".assoc.logistic")
haplogfile <- paste(output,sep="",".assoc.hap.logistic")

system(assoc)
system(logistic)
system(haplog)

f1 <- read.table(assocfile, header=T)
#print(f1)
f1 <- f1 %>% select(CHR,SNP,BP,F_A, F_U, OR, P) %>% rename(FREQ_AD=F_A, FREQ_CON=F_U, OR_assoc=OR, P_assoc=P)
print(f1,digits=2,row.names=F)
print("", quote=F)
#head(f1)

f1 <- f1 %>% select(CHR,SNP,BP,FREQ_AD,FREQ_CON)
f2 <- read.table(logisfile, header=T)
#print(f2)
f2 <- f2 %>% filter (TEST=="ADD") %>% select (SNP,NMISS,OR,L95,U95,P) %>% rename(OR_logis=OR, P_logis=P)
#head(f2)
f3<- left_join(f1,f2,by="SNP")
print(f3,digits=2,row.names=F)
print("", quote=F)
write.table(f3,out5,sep=",", quote=FALSE,row.names=F,col.names=T)

f4 <- read.table(haplogfile, header=T)
f4 <- f4 %>% select(-NSNP,-NHAP,-STAT) %>% arrange(desc(HAPLOTYPE))
print(f4,digits=2, row.names=F)
print("", quote=F)

hapomni <- paste("plink",sep=" ","--lfile", output, "--hap-window 4 --hap-logistic", "--covar", out3, "--covar-name RS,AUT,NCRAD,NW,ADSP,SEX,Age,APOE4dose,APOE2dose", "--ci 0.95", "--mhf 0.005 --hap-omnibus --out", output)
system(hapomni)
f4 <- read.table(haplogfile, header=T)
f4 <- f4 %>% rename(Pomni=P)
print(f4,digits=2, row.names=F)
print("", quote=F)

}
```

####Analyze SORCS2 dataset from Chris Medway
```{r,eval=F, echo=T}
cm2plink(input="SORCS2_CM.csv", output="SORCS2_CM")
```

####Analyze SORCS2 dataset from Ydb
```{r, eval=F, echo=T}
ydb2plink(input="SORCS2_Ydb.csv", output="SORCS2.Ydb")
#print(assoc)
```

####Analyze SORCS2 dataset from ADSP
```{r, eval=F, echo=T}
adsp2plink(input="SORCS2_ff.txt", output="SORCS2.adsp")
#print(assoc)
```

####Analyze combined (Ydb+ADSP) SORCS2 dataset
```{r, eval=F, echo=T}
comb2plink(output="SORCS2.comb")
#print(assoc)
```


###Joseph's Content
ADSP data was processed using the Genome GPS pipeline by the Bioinformatics Core in Rochestor. After variant calling (using 'GenotypeGVCFs' in the GATK pipeline), variants were annotated using BioR.  
Variants from genes of interest (here, SORCS2) were identified and processed using Perl scripts to generate files for downstream analysis.

```{r, eval=F}
system("perl generate_flat_file_SG.pl ADSP ADSP_sample_sex_age_AD_APOE.txt SORCS2_variants.txt 1000 SORCS2_ff.txt SORCS2_allele_counts.txt")

#Sex : 0: Male, 1: Female
#Case/Control: 0: Control, 1: AD, 2: Probable AD
#Call: 11/12/22
#Genotype: 0/0, 0/1, 1/1,  [./., ./0, ./1, 0/., 1/. encoded as NA]

#Importing Flat File
SORCS2_FF <- read.table("SORCS2_ff.txt",header = TRUE, check.names = FALSE)
head(SORCS2_FF)

#Importing allele counts for each variant
SORCS2_variants <- read.table("SORCS2_allele_counts.txt", header = TRUE, check.names = FALSE)
head(SORCS2_variants)
```






